x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java

```diff
@@ -264,7 +282,15 @@ public SegmentsStats segmentsStats(boolean includeSegmentFileSizes, boolean incl
         } else {
             return super.segmentsStats(includeSegmentFileSizes, includeUnloadedSegments);
         }
+    }
 
+    @Override
+    protected void closeNoLock(String reason, CountDownLatch closedLatch) {
+        super.closeNoLock(reason, closedLatch);
+        synchronized(closedListeners) {
+            IOUtils.closeWhileHandlingException(
+                closedListeners.stream().filter(Objects::nonNull).map(t -> (Closeable) () -> t.onClose(cacheIdentity))::iterator);
```

Comment:
Let's validate that closed listeners are not null when registering them instead of ignoring nulls here?

Reply:
done in eb04bdf7998
I had copied the above from Lucene's way of handling `readerClosedListeners` in `SegmentReader`.